### PR TITLE
Woche3

### PR DESCRIPTION
**Pull request by Koch-Dustin**
Changes:
1. plurality
2. runoff
3. tideman

### CODE REVIEW

File: Woche3/runoff/runoff.c [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            break;
+        }
+
+        eliminate(min);
+
+        for (int i = 0; i < candidate_count; i++) {
+            candidates[i].votes = 0;
+        }
+    }
+    return 0;
```
Comment (FrederikFricke) [inline Woche3/runoff/runoff.c:90]:
```suggestion
    }
    
    return 0;
```

File: Woche3/runoff/runoff.c [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+void eliminate(int min)
+{
+    for(int i = 0; i < candidate_count; i++) {
+        if(candidates[i].votes == min) {
+            candidates[i].eliminated = true;
+        }
+    }
+
+    return;
+}
```
Comment (FrederikFricke) [inline Woche3/runoff/runoff.c:163]: End of File

File: Woche3/runoff/runoff.c [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+           }
+        }
+    }
+
+    return;
+}
+
+bool print_winner(void) {
+    for (int i = 0; i < voter_count; i++) {  
+        int half_vote = voter_count / 2;
```
Comment (FrederikFricke) [inline Woche3/runoff/runoff.c:120]:
```suggestion
    int half_vote = voter_count / 2;
    for (int i = 0; i < voter_count; i++) {
```

File: Woche3/tideman/tideman.c [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+bool vote(int rank, string name, int ranks[]) {
+    
+    for (int i = 0; i < candidate_count; i++) {
+        if (strcmp(name,  candidates[i]) == 0) {
+            ranks[rank] = i;
+            return true;
+        }
+    }
+    return false;
```
Comment (FrederikFricke) [inline Woche3/tideman/tideman.c:90]:
```suggestion
    }
    
    return false;
```

File: Woche3/tideman/tideman.c [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+void record_preferences(int ranks[]) {
+
+    for(int i = 0; i < candidate_count; i++) {
+        for(int j = i + 1; j < candidate_count; j++) {
+            preferences[ranks[i]][ranks[j]] += 1;
+        }
+    }
+    return;
```
Comment (FrederikFricke) [inline Woche3/tideman/tideman.c:100]:
```suggestion
    }
    
    return;
```

File: Woche3/tideman/tideman.c [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    for (int i = 0; i < j; i++) {
+        if (rank[i] == true) {
+            rankpoint++;
+        }
+    }
+
+    if (rankpoint == j){
+        lock = false;
+    }
+    return false;
```
Comment (FrederikFricke) [inline Woche3/tideman/tideman.c:169]:
```suggestion
    }
    
    return false;
```

File: Woche3/tideman/tideman.c [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+void add_pairs(void) {
+    
+    for (int i = 0; i < candidate_count; i++) {
+        for (int j = i + 1; j < candidate_count; j++) {
+            if (preferences[i][j] > preferences[j][i]) {
+                pairs[pair_count].winner = i;
+                pairs[pair_count].loser = j;
+                pair_count++;
+            }
+            else if (preferences[i][j] < preferences[j][i]) {
```
Comment (FrederikFricke) [inline Woche3/tideman/tideman.c:112]:
```suggestion
            } else if (preferences[i][j] < preferences[j][i]) {
```

File: Woche3/tideman/tideman.c [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+            if (locked[k][i] == false) {
+                rank++;
+            }
+        }
+
+        if (rank == candidate_count) {
+            printf("%s\n", candidates[i]);
+        }
+    }
+}
```
Comment (FrederikFricke) [inline Woche3/tideman/tideman.c:203]: End of file

File: Woche3/runoff/runoff.c [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        if(candidates[i].eliminated == false && candidates[i].votes != voted_person_with_the_least_votes) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void eliminate(int voted_person_with_the_least_votes)
+{
```
Comment (FrederikFricke) [inline Woche3/runoff/runoff.c:156]:
```suggestion
void eliminate(int voted_person_with_the_least_votes) {
```

File: Woche3/tideman/tideman.c [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+        }
+    }
+
+    pair_count = 0;
+    int voter_count = get_int("Number of voters: ");
+
+    for (int i = 0; i < voter_count; i++) {
+
+        int ranks[candidate_count];
+
```
Comment (FrederikFricke) [inline Woche3/tideman/tideman.c:59]:
```suggestion
```

delete line

File: Woche3/tideman/tideman.c [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    add_pairs();
+    sort_pairs();
+    lock_pairs();
+    print_winner();
+    return 0;
+}
+
+bool vote(int rank, string name, int ranks[]) {
+
```
Comment (FrederikFricke) [inline Woche3/tideman/tideman.c:83]:
```suggestion
```

delete line

File: Woche3/tideman/tideman.c [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+        for(int j = i + 1; j < candidate_count; j++) {
+            preferences[ranks[i]][ranks[j]] += 1;
+        }
+    }
+
+    return;
+}
+
+void add_pairs(void) {
+
```
Comment (FrederikFricke) [inline Woche3/tideman/tideman.c:105]:
```suggestion
```

delete line

File: Woche3/tideman/tideman.c [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+                pairs[pair_count].winner = j;
+                pairs[pair_count].loser = i;
+                pair_count++;
+            }
+        }
+    }
+}
+
+void sort_pairs(void) {
+
```
Comment (FrederikFricke) [inline Woche3/tideman/tideman.c:122]:
```suggestion 
```

delete line

File: Woche3/tideman/tideman.c [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+        if (!lock) {
+            locked[pairs[i].winner][pairs[i].loser] = false;
+        }
+        lock = true;
+    }
+}
+
+void print_winner(void) {
+
```
Comment (FrederikFricke) [inline Woche3/tideman/tideman.c:187]:
```suggestion
```

delete line

**Review by FrederikFricke — CHANGES_REQUESTED** [review]

File: Woche3/runoff/runoff.c [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+        printf("Usage: runoff [candidate ...]\n");
+        return 1;
+    }
+
+    candidate_count = argc - 1;
+    if (candidate_count > MAX_CANDIDATES) {
+        printf("Maximum number of candidates is %i\n", MAX_CANDIDATES);
+        return 2;
+    }
+    for (int i = 0; i < candidate_count; i++) {
```
Comment (FrederikFricke) [inline Woche3/runoff/runoff.c:40]:
```suggestion
    }
    
    for (int i = 0; i < candidate_count; i++) {
```

File: Woche3/tideman/tideman.c [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+void lock_pairs(void) {
+    for (int i = 0; i < pair_count; i++) {
+        locked[pairs[i].winner][pairs[i].loser] = true;
+
+        validateLock(candidate_count);
+
+        if (!lock) {
+            locked[pairs[i].winner][pairs[i].loser] = false;
+        }
+        lock = true;
```
Comment (FrederikFricke) [inline Woche3/tideman/tideman.c:178]:
```suggestion
        }
        
        lock = true;
```

**Review by FrederikFricke — CHANGES_REQUESTED** [review]

**Review by FrederikFricke — APPROVED** [review]

File: Woche3/sort/answers.txt [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+
+How do you know?: TODO
+
+sort2 uses: TODO
+
+How do you know?: TODO
+
+sort3 uses: TODO
+
+How do you know?: TODO
```
Comment (stefanluth) [inline Woche3/sort/answers.txt:11]: Remove the file if it's not needed.

File: Woche3/plurality/plurality.c [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+            printf("Invalid vote.\n");
+        }
+    }
+
+    print_winner();
+}
+
+bool vote(string name) {
+    for (int i = 0; i < candidate_count; i++) {
+        if (strcmp(candidates[i].name, name) == 0) {
```
Comment (stefanluth) [inline Woche3/plurality/plurality.c:55]:
```suggestion
        bool variable = strcmp(candidates[i].name, name) == 0;
        if (variable) {
```

Use variables for more complex if-conditions or when it's not obvious what the condition does. 

Also do this in the other files.

File: Woche3/runoff/runoff.c [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+
+int voter_count;
+int candidate_count;
+
+bool vote(int voter, int rank, string name);
+void tabulate(void);
+bool print_winner(void);
+int find_min(void);
+bool is_tie(int min);
+void eliminate(int min);
```
Comment (stefanluth) [inline Woche3/runoff/runoff.c:27]: Sort by type and then alphabetically.

Comment (stefanluth) [reply Woche3/runoff/runoff.c:27]: In other files as well.

File: Woche3/tideman/tideman.c [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+        record_preferences(ranks);
+
+        printf("\n");
+    }
+
+    add_pairs();
+    sort_pairs();
+    lock_pairs();
+    print_winner();
+    return 0;
```
Comment (stefanluth) [inline Woche3/tideman/tideman.c:78]:
```suggestion
    print_winner();

    return 0;
```

File: Woche3/tideman/tideman.c [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+        for (int k = 0; k < candidate_count; k++) {
+            if (locked[k][i] == false) {
+                rank++;
+            }
+        }
+
+        if (rank == candidate_count) {
+            printf("%s\n", candidates[i]);
+        }
+    }
```
Comment (stefanluth) [inline Woche3/tideman/tideman.c:198]:
```suggestion
    }
    
    return;
```

File: Woche3/tideman/tideman.c [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+                pairs[pair_count].winner = j;
+                pairs[pair_count].loser = i;
+                pair_count++;
+            }
+        }
+    }
+}
+
+void sort_pairs(void) {
+    for(int i = 0; i < pair_count; i++) {
```
Comment (stefanluth) [inline Woche3/tideman/tideman.c:119]:
```suggestion
    for (int i = 0; i < pair_count; i++) {
```

File: Woche3/tideman/tideman.c [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+        int max = i;
+        for(int j = i + 1; j < pair_count; j++) {
```
Comment (stefanluth) [inline Woche3/tideman/tideman.c:121]:
```suggestion
        for (int j = i + 1; j < pair_count; j++) {
```

File: Woche3/tideman/tideman.c [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+            bool preference_votes_are_lower_than_maximum_votes = preferences[pairs[j].winner][pairs[j].loser] > preferences[pairs[max].winner][pairs[max].loser];
+            if(preference_votes_are_lower_than_maximum_votes) {
```
Comment (stefanluth) [inline Woche3/tideman/tideman.c:123]:
```suggestion
            if (preference_votes_are_lower_than_maximum_votes) {
```

In other files as well.

**Review by stefanluth — CHANGES_REQUESTED** [review]

File: Woche3/sort/answers.txt (hunk 17 above)
Comment (stefanluth) [reply Woche3/sort/answers.txt:11]: The other files in `./sort` as well

File: Woche3/tideman/tideman.c [hunk 25]
```diff
@@ -0,0 +1,202 @@
...
+                pairs[pair_count].loser = j;
+                pair_count++;
+            } else if (preferences[i][j] < preferences[j][i]) {
+                pairs[pair_count].winner = j;
+                pairs[pair_count].loser = i;
+                pair_count++;
+            }
+        }
+    }
+}
```
Comment (stefanluth) [inline Woche3/tideman/tideman.c:117]:
```suggestion
    }
    
    return;
}
```

Comment (stefanluth) [reply Woche3/tideman/tideman.c:117]: In other functions & files as well.

**Review by stefanluth — CHANGES_REQUESTED** [review]
Delete `Woche3/.DS_Store`

File: Woche3/plurality/plurality.c [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+    candidate swap;
+    
+    for (int i = 0, n = candidate_count; i < n; i++) {
+        if (candidates[i].votes > candidates[i + 1].votes) {
+            swap = candidates[i];
+            candidates[i] = candidates[i + 1]; 
+            candidates[i + 1] = swap;
+        }
+
+        if(candidates[i].votes == candidates[i + 1].votes) {
```
Comment (stefanluth) [inline Woche3/plurality/plurality.c:75]:
```suggestion
        if (candidates[i].votes == candidates[i + 1].votes) {
```

Comment (stefanluth) [reply Woche3/plurality/plurality.c:75]: In other places as well.

File: Woche3/runoff/runoff.c [hunk 27]
```diff
@@ -0,0 +1,164 @@
...
+            candidates[i].votes = 0;
+        }
+    }
+
+    return 0;
+}
+
+bool vote(int voter, int rank, string name) {
+    for (int i = 0; i < candidate_count; i++) {
+        if (strcmp(name, candidates[i].name) == 0) {
```
Comment (stefanluth) [inline Woche3/runoff/runoff.c:97]: Extract complex or unclear if-conditions to a variable.

Comment (stefanluth) [reply Woche3/runoff/runoff.c:97]: In other places as well.

**Review by stefanluth — CHANGES_REQUESTED** [review]
Same remarks as my first review.

**Review by DevBastiS — APPROVED** [review]

**Review by stefanluth — CHANGES_REQUESTED** [review]
Still unresolved.

**Review by stefanluth — CHANGES_REQUESTED** [review]
Delete `Woche3/.DS_Store`

File: Woche3/plurality/plurality.c (hunk 26 above)
Comment (stefanluth) [reply Woche3/plurality/plurality.c]: > In other places as well.

Still unresolved.

**Comment by stefanluth** [comment]
https://github.com/Koch-Dustin/Cs50Kurs/pull/5#discussion_r955013350
https://github.com/Koch-Dustin/Cs50Kurs/pull/5#discussion_r955013543

Bitte erst resolven

File: Woche3/runoff/runoff.c [hunk 28]
```diff
@@ -0,0 +1,164 @@
...
+            minimal_vote = candidates[i].votes;
+        }
+    }
+
+    return minimal_vote;
+}
+
+bool is_tie(int voted_person_with_the_least_votes) {
+    for(int i = 0; i < candidate_count; i++) {
+        if(candidates[i].eliminated == false && candidates[i].votes != voted_person_with_the_least_votes) {
```
Comment (stefanluth) [inline Woche3/runoff/runoff.c:148]:
```suggestion
        if (candidates[i].eliminated == false && candidates[i].votes != voted_person_with_the_least_votes) {
```

File: Woche3/runoff/runoff.c [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void eliminate(int voted_person_with_the_least_votes) {
+    for(int i = 0; i < candidate_count; i++) {
+        if(candidates[i].votes == voted_person_with_the_least_votes) {
```
Comment (stefanluth) [inline Woche3/runoff/runoff.c:158]:
```suggestion
        if (candidates[i].votes == voted_person_with_the_least_votes) {
```

File: Woche3/runoff/runoff.c (hunk 28 above)
Comment (stefanluth) [reply Woche3/runoff/runoff.c:148]: Extract if-condition to variable.

File: Woche3/tideman/tideman.c [hunk 30]
```diff
@@ -0,0 +1,203 @@
...
+    sort_pairs();
+    lock_pairs();
+    print_winner();
+
+    return 0;
+}
+
+bool vote(int rank, string name, int ranks[]) {
+    for (int i = 0; i < candidate_count; i++) {
+        if (strcmp(name,  candidates[i]) == 0) {
```
Comment (stefanluth) [inline Woche3/tideman/tideman.c:83]: Extract if-condition to variable.

File: Woche3/tideman/tideman.c [hunk 31]
```diff
@@ -0,0 +1,203 @@
...
+        }
+    }
+
+    return;
+}
+
+void add_pairs(void) {
+    for (int i = 0; i < candidate_count; i++) {
+        for (int j = i + 1; j < candidate_count; j++) {
+            if (preferences[i][j] > preferences[j][i]) {
```
Comment (stefanluth) [inline Woche3/tideman/tideman.c:105]: Extract if-condition to variable.

File: Woche3/tideman/tideman.c [hunk 32]
```diff
@@ -0,0 +1,203 @@
...
+                pairs[pair_count].winner = i;
+                pairs[pair_count].loser = j;
+                pair_count++;
+            } else if (preferences[i][j] < preferences[j][i]) {
```
Comment (stefanluth) [inline Woche3/tideman/tideman.c:109]: Extract if-condition to variable.

File: Woche3/tideman/tideman.c [hunk 33]
```diff
@@ -0,0 +1,206 @@
...
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void record_preferences(int ranks[]) {
+    for(int i = 0; i < candidate_count; i++) {
+        for(int j = i + 1; j < candidate_count; j++) {
```
Comment (stefanluth) [inline Woche3/tideman/tideman.c:95]:
```suggestion
    for (int i = 0; i < candidate_count; i++) {
        for (int j = i + 1; j < candidate_count; j++) {
```

**Review by stefanluth — APPROVED** [review]

**Review by shoreal — APPROVED** [review]

**Pull request merged**